### ISSUE TITLE
Questions about your algorithm Reg-CFR

### ISSUE THREAD

**Issue by sakura970** [liumy2010/LiteEFG#6]
hello, mingyang,

I am following your work Reg-CFR, but I find that a perturbation weight $\gamma$ is used in your paper, but I find that your implementation of Reg-CFR does not have this item. 

In addition, I would like to understand how to calculate the exploitability of a strategy profile in the perturbed game based on your code.

**Comment by liumy2010** [comment]
1. I've just added `gamma` to Reg-CFR in the latest version. Thank you for your reminder!
2. Now LiteEFG allows `Environment.utility(strategy_list, type_name="default")`. Therefore, you may use `Environment.utility([best_response_strategy_in_perturbed_game, current_strategy])` to compute the exploitability of player 2's `current_strategy`. `best_response_strategy_in_perturbed_game` can be computed via mixing the best-response and a uniform strategy in each infoset.

**Comment by sakura970** [comment]
Thank you very much for your reply, can you provide details on how to calculate ``best_response_strategy_in_perturbed_game`` and mix strategies. Because I didn't find how to calculate ``best_response_strategy_in_perturbed_game`` and mix strategies in the existing code.

**Comment by liumy2010** [comment]
```python
best_response_strategy = LiteEFG.argmax(counterfactual_value)
best_response_strategy_in_perturbed_game = (1.0-gamma) * best_response_strategy + gamma / self.action_set_size
```

**Comment by sakura970** [comment]
Thanks for your reply, but can you provide the implementation of such a class, because I don't know how to input a strategy into such a class to calculate the counterfactual value.

Also, regarding your funtion ``env.utility``, can you give some examples so that I can understand how to use it.

**Comment by liumy2010** [comment]
To compute `counterfactual_values`, `baselines/CFR.py` may serve as an example. In `README.md`, you can find how to use `Environment.exploitability` and `Environment.utility`. An example can be found in `baselines/utils.py`.

**Comment by sakura970** [comment]
Sorry, I think you misunderstood what I meant. What I mean is that given a strategy, how to calculate its counterfactual value at each information set? Your code ``baselines/CFR.py`` only supports calculations for uniform strategies.

Additionally, I find that ``utils.py`` does not seem to have ``env.utility``.

**Comment by sakura970** [comment]
Here is my code
```
import pyspiel
import LiteEFG


class CFRPlusSolver(LiteEFG.Graph):
    def __init__(self, gamma=1, true_last_iterate_convergence=False):
        self.true_last_iterate_convergence = true_last_iterate_convergence
        super().__init__()
        
        with LiteEFG.backward(is_static=True):
                
            expectation = LiteEFG.const(1, 0.0)
            self.strategy = LiteEFG.const(self.action_set_size, 1.0 / self.action_set_size)
            self.regret_buffer = LiteEFG.const(self.action_set_size, 0.0)

            self.avg_seq_strategy = LiteEFG.const(self.action_set_size, 0.0)
            self.avg_strategy = self.strategy.copy()
            self.iteration = LiteEFG.const(1, 0)
            self.strategy_coef = LiteEFG.const(1, 1)

        # RM+
        with LiteEFG.backward():
            self.iteration.inplace(self.iteration+1)
            self.strategy_coef.inplace(self.iteration ** gamma)

            counterfactual_value = LiteEFG.aggregate(expectation, aggregator="sum") + self.utility
            expectation.inplace(LiteEFG.dot(counterfactual_value, self.strategy))
            self.regret_buffer.inplace(LiteEFG.maximum(self.regret_buffer + counterfactual_value - expectation, 0.0))
            self.strategy.inplace(LiteEFG.normalize(self.regret_buffer, p_norm=1.0, ignore_negative=True))

            self.avg_seq_strategy.inplace(self.avg_seq_strategy + self.strategy_coef * self.strategy * self.reach_prob)
            self.avg_strategy.inplace(self.avg_seq_strategy.normalize(p_norm=1.0, ignore_negative=True))

        print("===============Graph is ready for CFR+===============")
        print()
        print("====================================================\n")
    
    def update_graph(self, env : LiteEFG.Environment) -> None:
        env.update(self.strategy, upd_player=1)
        env.update(self.strategy, upd_player=2)
    
    def current_strategy(self, type_name="last-iterate") -> LiteEFG.GraphNode:
        assert(type_name in ["last-iterate", "linear-avg-iterate"])
        if not self.true_last_iterate_convergence:
            return self.strategy if type_name == "linear-avg-iterate" else self.avg_strategy 
        else:
            return self.strategy
   

class PerturbedGameBestResponseSolver(LiteEFG.Graph):
    def __init__(self, strategy, perturbed_gamma = 1e-2):
        super().__init__()
        # self.strategy = strategy.copy()
        with LiteEFG.backward(is_static=True):

            expectation = LiteEFG.const(size=1, val=0.0)
            self.strategy = strategy.copy()#LiteEFG.const(self.action_set_size, 1.0 / self.action_set_size)
            self.regret_buffer = LiteEFG.const(self.action_set_size, 0.0)
            self.perturbed_strategy = LiteEFG.const(self.action_set_size, perturbed_gamma)
            self.best_response_strategy_in_perturbed_game = LiteEFG.const(self.action_set_size, 1.0 / self.action_set_size)

        with LiteEFG.backward():
            counterfactual_value = LiteEFG.aggregate(expectation, aggregator="sum") + self.utility
            
            best_response_strategy = LiteEFG.argmax(counterfactual_value)
            weight = 1 - perturbed_gamma*self.action_set_size
            self.best_response_strategy_in_perturbed_game = weight * best_response_strategy + self.perturbed_strategy

        print("===============Graph is ready for PerturbedGameBestResponseSolver===============")
        print()
        print("====================================================\n")
    
    def update_graph(self, env : LiteEFG.Environment) -> None:
        env.update(self.strategy)
    
    def current_strategy(self, type_name="last-iterate") -> LiteEFG.GraphNode:
        return self.best_response_strategy_in_perturbed_game
    

solver = CFRPlusSolver()
game_env='kuhn_poker'
game = pyspiel.load_game(game_env)
env = LiteEFG.OpenSpielEnv(game, regenerate=False, traverse_type="Enumerate")
env.set_graph(solver)

for i in range(100):
    solver.update_graph(env)
    convergence_type = 'last-iterate'
    env.update_strategy(solver.current_strategy(), update_best=(convergence_type == "best-iterate"))

# env = LiteEFG.OpenSpielEnv(game, regenerate=False, traverse_type="Enumerate")
PerturbedGameBestResponsesolver = PerturbedGameBestResponseSolver(perturbed_gamma = 1e-2, strategy=solver.current_strategy())
env.set_graph(PerturbedGameBestResponsesolver)

PerturbedGameBestResponsesolver.update_graph(env)
convergence_type = 'last-iterate'
env.update_strategy(PerturbedGameBestResponsesolver.current_strategy(), update_best=(convergence_type == "best-iterate"))

print(env.utility([PerturbedGameBestResponsesolver.current_strategy(), solver.current_strategy()]))
```
However, I encounted the following error
```
Traceback (most recent call last):
  File "XXXX.py", line 74, in <module>
    env.set_graph(PerturbedGameBestResponsesolver)
ValueError: The inputs of ConstVector must be scalars
```

**Comment by liumy2010** [comment]
1. For the code, I do not encounter the error above. May you check whether your LiteEFG is the latest version?
2. However, the problem of the code is `self.strategy = strategy.copy()` in `PerturbedGameBestResponseSolver`. When calling `env.set_graph`, the previous graph will be completely removed. Therefore, you may not use `strategy` defined in the previous computation graph to define a new computation graph. Instead, please define all computation in the same class, but with different colors. You may check `baselines/Reg_DOMD.py` for details.
3. `baselines/CFR.py` computes the counterfactual value corresponding to the current strategy. Please check its explanation in `README.md` or the LiteEFG paper.
4. `env.utility` has the same grammar as `env.exploitability`.

**Comment by sakura970** [comment]
Regarding your 2 and 3, this is exactly what I want to say. Your current code cannot calculate the counterfactual value for any given strategy. Your code can only compute the counterfactual value for certain specific strategies. For example, for any strategy that is neither the current strategy nor the uniform strategy, your code cannot calculate the counterfactual value for it at all.

Regarding your 4, I think I find the problem. It seems that ``env.utility`` only supports the strategy of a single graph output for calculation. If I want to test the strategies that different graph outputs, it seems to be unfeasible. Is there any way to solve this problem?

**Comment by sakura970** [comment]
In other words, can you provide a method to convert the strategy output by graph into the policy defined in openspiel？Therefore, I can bypass this limitation through the policy defined in openspiel.

**Comment by sakura970** [comment]
I just found that the usage example of ``env.exploitability`` in your code ``baselines/utils.py`` has changed relative to version 0.1.2. Could you explain what the return value of ``env.exploitability`` in the latest version is?

From your code ``baselines/utils.py``, I guess that ``env.exploitability`` currently outputs the exploitability based on ``convergence_type`` and the lowest exploitability of all iterates. However, this does not seem to be the case with our code.

```
import pyspiel
import LiteEFG


class CFRPlusSolver(LiteEFG.Graph):
    def __init__(self, gamma=1, true_last_iterate_convergence=False):
        self.true_last_iterate_convergence = true_last_iterate_convergence
        super().__init__()
        
        with LiteEFG.backward(is_static=True):
                
            expectation = LiteEFG.const(1, 0.0)
            self.strategy = LiteEFG.const(self.action_set_size, 1.0 / self.action_set_size)
            self.regret_buffer = LiteEFG.const(self.action_set_size, 0.0)

            self.avg_seq_strategy = LiteEFG.const(self.action_set_size, 0.0)
            self.avg_strategy = self.strategy.copy()
            self.iteration = LiteEFG.const(1, 0)
            self.strategy_coef = LiteEFG.const(1, 1)
            self.reach_probs = LiteEFG.const(self.action_set_size, 0.0)

        # RM+
        with LiteEFG.backward():
            self.iteration.inplace(self.iteration+1)
            self.strategy_coef.inplace(self.iteration ** gamma)

            counterfactual_value = LiteEFG.aggregate(expectation, aggregator="sum") + self.utility
            expectation.inplace(LiteEFG.dot(counterfactual_value, self.strategy))
            self.regret_buffer.inplace(LiteEFG.maximum(self.regret_buffer + counterfactual_value - expectation, 0.0))
            self.strategy.inplace(LiteEFG.normalize(self.regret_buffer, p_norm=1.0, ignore_negative=True))

            self.avg_seq_strategy.inplace(self.avg_seq_strategy + self.strategy_coef * self.strategy * self.reach_prob)
            self.avg_strategy.inplace(self.avg_seq_strategy.normalize(p_norm=1.0, ignore_negative=True))
            self.reach_probs.inplace(LiteEFG.const(self.action_set_size, self.reach_prob))

        print("===============Graph is ready for CFR+===============")
        print()
        print("====================================================\n")
    
    def update_graph(self, env : LiteEFG.Environment) -> None:
        env.update(self.strategy, upd_player=1)
        env.update(self.strategy, upd_player=2)
    
    def current_strategy(self, type_name="last-iterate") -> LiteEFG.GraphNode:
        if not self.true_last_iterate_convergence:
            return self.strategy if type_name == "linear-avg-iterate" else self.avg_strategy 
        else:
            return self.strategy




solver = CFRPlusSolver(true_last_iterate_convergence=True)
game_env='kuhn_poker'
game = pyspiel.load_game(game_env)
env = LiteEFG.OpenSpielEnv(game, regenerate=False, traverse_type="Enumerate")
env.set_graph(solver)

for i in range(1000):
    solver.update_graph(env)
    convergence_type = 'last-iterate'
    env.update_strategy(solver.current_strategy(), update_best=(convergence_type == "best-iterate"))


print(env.exploitability(solver.current_strategy(), convergence_type))
```

```
===============Graph is ready for CFR+===============

====================================================

[0.007015969441116605, 0.031847777802421415]
```

**Comment by sakura970** [comment]
I just found that the usage example of ``env.exploitability`` in your code ``baselines/utils.py`` has been recovered to 0.1.2. Could you explain which one is correct? 

Also, could you reply my following comments? I've been waiting for you for almost a week.

>Regarding your 2 and 3, this is exactly what I want to say. Your current code cannot calculate the counterfactual value for any given strategy. Your code can only compute the counterfactual value for certain specific strategies. For example, for any strategy that is neither the current strategy nor the uniform strategy, your code cannot calculate the counterfactual value for it at all.

>Regarding your 4, I think I find the problem. It seems that env.utility only supports the strategy of a single graph output for calculation. If I want to test the strategies that different graph outputs, it seems to be unfeasible. Is there any way to solve this problem?

**Comment by liumy2010** [comment]
The latest version of `utility.py` is correct, and it aligns with the description in `README`.

- There might be some misunderstanding about how `CFR.py` works. It would be helpful to read the explanation of `CFR.py` in `README` or the paper of LiteEFG.
- Yes. `env.utility` only supports the strategy in a single graph, because the environment only supports loading a single graph simultaneously. Therefore, if you want to load multiple graphs, you may need to use the `color` arguments to merge them into one graph. `Reg_DOMD.py` may be helpful since it merges the DOMD update rule and the annealing of $\tau$ into a single computation graph.

**Comment by sakura970** [comment]
Thanks for your response. However, I still have the following questions.

> There might be some misunderstanding about how CFR.py works. It would be helpful to read the explanation of CFR.py in README or the paper of LiteEFG.

I cannot understand this. For any strategy represented by the OpenSpiel Policy class, how to calculate the counterfactual value? For example, how to initialize this given strategy in the ``graph`` class? Your code only supports for uniform strategy or pure strategy. If possible, I hope you can provide the code instead of asking me to read ``README``.

> Therefore, if you want to load multiple graphs, you may need to use the color arguments to merge them into one graph. Reg_DOMD.py may be helpful since it merges the DOMD update rule and the annealing of τ into a single computation graph.

I don't agree with this response. Given any two strategies represented by the OpenSpiel Policy class, how can I merge them into the same graph?

In summary, my main question is, given any strategy represented by the OpenSpiel Policy class, how to initialize it in the ``graph`` class? Your code is

```
with LiteEFG.backward(is_static=True):

            expectation = LiteEFG.const(size=1, val=0.0)
            self.strategy = LiteEFG.const(self.action_set_size, 1.0 / self.action_set_size)
            self.regret_buffer = LiteEFG.const(self.action_set_size, 0.0)
```

``self.strategy`` can only be uniform strategy or pure strategy because there does not exist any ``key value`` operator .
